File: 1016.제곱ㄴㄴ수/solve.py
def solve(n, m):
    sieve = [1] * (m - n + 1)
    for k in range(2, int(m ** 0.5) + 1):
        s = k * k
        i = (s - (n % s)) % s
        for j in range(i, len(sieve), s):
            sieve[j] = 0
    return sum(sieve)
    
n, m = map(int, input().split())
print(solve(n, m))

File: 1016.제곱ㄴㄴ수/solve.cpp
#include <bits/stdc++.h>
using namespace std;

typedef unsigned long long long_t;

int solve(long_t n, long_t m) {
    vector<bool> sieve(m - n + 1, true);
    for (long_t k = 2; k * k <= m; k++) {
        long_t s = k * k;
        long_t i = (s - (n % s)) % s;
        for (long_t j = i; j < sieve.size(); j += s)
            sieve[j] = false;
    }
    int cnt = 0; 
    for (int i = 0; i < sieve.size(); i++)
        if (sieve[i]) cnt++;
    return cnt;
}

int main() {
    long_t n, m; cin >> n >> m;
    cout << solve(n, m);
}